src/x11/x11-platform.c

```diff
@@ -1311,34 +1315,73 @@ static EGLBoolean eplX11WaitGL(EplDisplay *pdpy, EplSurface *psurf)
     return ret;
 }
 
-EGLAttrib *eplX11GetInternalSurfaceAttribs(EplPlatformData *plat, EplDisplay *pdpy, const EGLAttrib *attribs)
+EGLAttrib *eplX11GetInternalSurfaceAttribs(EplPlatformData *plat,
+        EplDisplay *pdpy, EplSurfaceType surface_type, const EGLAttrib *attribs)
 {
     EGLAttrib *internalAttribs = NULL;
-    int count = 0;
+    int count = eplCountAttribs(attribs);
+
+    internalAttribs = malloc((count + 3) * sizeof(EGLAttrib));
+    if (internalAttribs == NULL)
+    {
+        eplSetError(plat, EGL_BAD_ALLOC, "Out of memory\n");
+        return NULL;
+    }
+
+    count = 0;
+    internalAttribs[count++] = EGL_SURFACE_Y_INVERTED_NVX;
+    internalAttribs[count++] = EGL_TRUE;
 
     if (attribs != NULL)
     {
-        for (count = 0; attribs[count] != EGL_NONE; count += 2)
+        int i;
+        for (i=0; attribs[i] != EGL_NONE; i += 2)
         {
-            if (attribs[count] == EGL_SURFACE_Y_INVERTED_NVX)
+            if (attribs[i] == EGL_SURFACE_Y_INVERTED_NVX)
             {
-                eplSetError(plat, EGL_BAD_ATTRIBUTE, "Invalid attribute 0x%04x\n", attribs[count]);
+                eplSetError(plat, EGL_BAD_ATTRIBUTE, "Invalid attribute 0x%04x\n", attribs[i]);
+                free(internalAttribs);
                 return NULL;
             }
+            else if (attribs[i] == EGL_RENDER_BUFFER)
+            {
+                if (surface_type != EPL_SURFACE_TYPE_WINDOW)
+                {
+                    /*
+                     * The EGL_RENDER_BUFFER is not valid for pixmaps, since
+                     * they're always single-buffered.
+                     */
+                    eplSetError(plat, EGL_BAD_ATTRIBUTE, "EGL_RENDER_BUFFER is not valid for pixmaps");
+                    free(internalAttribs);
+                    return NULL;
+                }
+                else if (attribs[i + 1] == EGL_SINGLE_BUFFER)
+                {
+                    /*
+                     * We don't currently support single-buffered rendering for
+                     * a window, but this attribute is only a hint, so it's not
+                     * an error to request it.
+                     */
+                    plat->callbacks.debugMessage(EGL_DEBUG_MSG_WARN_KHR,
+                            "EGL_SINGLE_BUFFER requested, but single-buffered rendering is not supported");
+                }
+                else if (attribs[i + 1] != EGL_BACK_BUFFER)
+                {
+                    eplSetError(plat, EGL_BAD_ATTRIBUTE,
+                            "Invalid EGL_RENDER_BUFFER value 0x%04x", attribs[i + 1]);
+                    free(internalAttribs);
+                    return NULL;
+                }
+            }
```

Comment:
Just to check I'm understanding correctly, we don't have to copy the `EGL_RENDER_BUFFER` to the internal attribute list because we will always know the value of it at query time due to the type of surface being used right? i.e. we always know pbuffer is single and window is double. I might also request a small comment about that here since it took a minute for me to realize that.